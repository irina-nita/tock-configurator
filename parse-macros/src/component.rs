// Copyright OxidOS Automotive 2024.

use darling::ast::NestedMeta;
use darling::{util::Flag, FromMeta};
use proc_macro::TokenStream;
use proc_macro2::Literal;
use quote::{format_ident, quote, ToTokens};
use syn::ItemStruct;

/// `peripheral` attribute proc macro arguments.
#[derive(Debug, FromMeta)]
struct PeripheralMacroArgs {
    peripherals: Option<String>,
    #[darling(flatten)]
    component_args: ComponentMacroArgs,
}

/// `component` attribute proc macro arguments.
#[derive(Debug, FromMeta)]
struct ComponentMacroArgs {
    serde: Flag,
    curr: Flag,
    ident: String,
}

/// Helper function used after parsing the macro attributes for generating the struct and the impl
/// blocks for the component.
fn define_struct(
    base_ident: String,
    curr: Flag,
    serde: Flag,
    peripherals: Option<String>,
    item: TokenStream,
) -> syn::Result<TokenStream> {
    // Get the struct definition.
    let struct_item = syn::parse::<ItemStruct>(item)?;

    // Get the field that matches the field specified by the attribute.
    let fields = struct_item
        .fields
        .iter()
        .map(|f| f.to_token_stream())
        .collect::<Vec<_>>();

    // Retrieve information from the struct for redefining.
    let struct_ty = struct_item.ident;
    let struct_generics = struct_item.generics;

    let mut arg_num: usize = 0;
    let mut tuple_struct = false;

    let derive_default = if struct_item.fields.is_empty() {
        Some(quote!( #[derive(Default)] ))
    } else {
        None
    };

    // Get the identifiers and the types of the fields for the initialization function.
    let (field_names, field_types): (Vec<_>, Vec<_>) = struct_item
        .fields
        .iter()
        .map(|f| {
            match f.ident.clone() {
                Some(ident) => (ident.to_token_stream(), f.ty.clone()),
                None => {
                    let field = (
                        format_ident!("arg{}", arg_num.to_string()).to_token_stream(),
                        f.ty.clone(),
                    );
                    // This should be set only once but it's fine for now.
                    tuple_struct = true;
                    arg_num += 1;
                    field
                }
            }
        })
        .unzip();

    let struct_vis = struct_item.vis;

    // The crate's name is based on the `curr` flag.
    let _crate = if curr.is_present() {
        quote!(crate)
    } else {
        quote!(parse)
    };

    let default_fn = _crate.to_string() + "::Uuid::new_v4";

    let (derive_serde, serde_skip) = serde
        .is_present()
        .then_some((
            quote!(#[derive(serde::Serialize, serde::Deserialize)]),
            quote!(#[serde(skip, default = #default_fn)]),
        ))
        .unzip();

    // Get generics and all.
    let (impl_generics, ty_generics, where_clause) = struct_generics.split_for_impl();

    let peripherals = peripherals.map(|p_ident| format_ident!("{}", p_ident.clone()));

    let arg_num_ident = Literal::usize_unsuffixed(arg_num);

    let (init_expr, self_ident, struct_ast) = match (tuple_struct, peripherals) {
        (true, None) => (
            quote! (Self (#(#field_names,)* #_crate::Uuid::new_v4())),
            quote! {
                use #_crate::FormatIdent;
                Ok(String::from(#base_ident) + &self.#arg_num_ident.format_ident())
            },
            quote! {
                #derive_default
                #derive_serde
                #struct_vis struct #struct_ty #struct_generics (#(#field_types,)*
                #serde_skip
                #_crate::Uuid);
            },
        ),
        (true, Some(p_ident)) => (
            quote! (Self (#(#field_names,)*)),
            quote! {
                Ok(#p_ident.clone() + &String::from(#base_ident))
            },
            quote! {
                #derive_default
                #derive_serde
                #struct_vis struct #struct_ty #struct_generics (#(#field_types,)*);
            },
        ),
        (false, None) => (
            quote! {Self {
                 #(#field_names,)*
                __ident: #_crate::Uuid::new_v4()
            }
            },
            quote! {
                use #_crate::FormatIdent;
                Ok(String::from(#base_ident) + &self.__ident.format_ident())
            },
            quote! {
            #derive_default
            #derive_serde
            #struct_vis struct #struct_ty #struct_generics {
                #(#fields,)*
                #serde_skip
                __ident: #_crate::Uuid,
            }
            },
        ),
        (false, Some(p_ident)) => (
            quote! {Self {
                 #(#field_names,)*
            }
            },
            quote! {
                Ok(#p_ident.clone() + &String::from(#base_ident))
            },
            quote! {
            #derive_default
            #derive_serde
            #struct_vis struct #struct_ty #struct_generics {
                #(#fields,)*
            }
            },
        ),
    };

    // The final code generated with both the constructor and the `Ident` trait implementations.
    let impls = quote! {
        impl #impl_generics #struct_ty #ty_generics #where_clause {
            pub fn new(#(#field_names : #field_types),*) -> Self {
                #init_expr
            }
        }

        impl #impl_generics #_crate::component::Ident for #struct_ty #ty_generics #where_clause {
            fn ident(&self) -> Result<String, #_crate::error::Error> {
                #self_ident
            }
        }
    };

    // The code generated with the constructor.
    Ok(quote!(
        #struct_ast
        #impls
    )
    .into())
}

/// Generate the expression for defining a component which is also a peripheral.
pub(crate) fn define_peripheral(attrs: TokenStream, item: TokenStream) -> TokenStream {
    // Get the dependency field from the attribute.
    let attr_args = match NestedMeta::parse_meta_list(attrs.into()) {
        Ok(v) => v,
        Err(err) => {
            return TokenStream::from(darling::Error::from(err).write_errors());
        }
    };

    let args = match PeripheralMacroArgs::from_list(&attr_args) {
        Ok(v) => v,
        Err(err) => {
            return TokenStream::from(err.write_errors());
        }
    };

    let peripherals_ident = args.peripherals.unwrap_or("PERIPHERALS".to_string());

    match define_struct(
        args.component_args.ident,
        args.component_args.curr,
        args.component_args.serde,
        Some(peripherals_ident),
        item,
    ) {
        Ok(tokens) => tokens,
        Err(err) => err.into_compile_error().into(),
    }
}

/// Generate the expression for defining a component.
pub(crate) fn define_component(attrs: TokenStream, item: TokenStream) -> TokenStream {
    // Get the dependency field from the attribute.
    let attr_args = match NestedMeta::parse_meta_list(attrs.into()) {
        Ok(v) => v,
        Err(err) => {
            return TokenStream::from(darling::Error::from(err).write_errors());
        }
    };

    let args = match PeripheralMacroArgs::from_list(&attr_args) {
        Ok(v) => v,
        Err(err) => {
            return TokenStream::from(err.write_errors());
        }
    };

    match define_struct(
        args.component_args.ident,
        args.component_args.curr,
        args.component_args.serde,
        None,
        item,
    ) {
        Ok(tokens) => tokens,
        Err(err) => err.into_compile_error().into(),
    }
}
