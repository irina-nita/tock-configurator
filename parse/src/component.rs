// Copyright OxidOS Automotive 2024.

use std::any::{Any, TypeId};
use std::rc::Rc;

/// Constants used in the identifying process for component nodes that
/// are sure to never be instantiated for the same type twice.
pub mod constants {
    use crate::static_ident;
    pub use once_cell::sync::Lazy;

    pub static PERIPHERALS: Lazy<String> = static_ident!("peripherals");
    pub static CHIP: Lazy<String> = static_ident!("chip");
}

/// A trait for objects that define a Tock `capsule`.
///
/// Besides the [`crate::Component`] implementation, these types must provide the code
/// expression that returns the driver number for the capsule.
///
/// # Example
///
/// ```rust,ignore
/// use quote::quote;
///
/// impl Capsule for Console {
///     fn driver_num() -> proc_macro2::TokenStream {
///         quote!(capsules::core::console::DRIVER_NUM);
///     }
/// }
/// ```
pub trait Capsule: crate::Component {
    fn driver_num(&self) -> proc_macro2::TokenStream;
}

/// A trait for objects that define variables represented by an **unique** identifier.
///
/// Additionally, the procedural macro [`parse_macros::component`] provides an implementation
/// of the [`Ident`] trait.
///
/// # Example
///
/// ```rust,ignore
/// // Proc macro that implements the `Ident` trait.
/// #[parse::node(ident = "console")]
/// struct Console {
///     /* ... */
/// }
/// ```
pub trait Ident {
    fn ident(&self) -> Result<String, crate::error::Error>;
}

//  TODO: The ident trait must have a static str? so that this is not to be confused?
// This can't be object safe unless... the ident parameter is an uuid (has default),
// and the base is a static/ implemented in the ident function?

impl Ident for super::NoSupport {
    fn ident(&self) -> Result<String, crate::error::Error> {
        Err(crate::error::Error::CodeNotProvided)
    }
}

impl<P: crate::DefaultPeripherals> Ident for P {
    fn ident(&self) -> Result<String, crate::error::Error> {
        Ok(constants::PERIPHERALS.clone())
    }
}

/// A trait for objects that define variables from a function in the platform's main.
///
/// Besides the [`Ident`] implementation, the *component* can optionally provide the type of the
/// variable, the code expression that returns a new instance, and the dependencies for the
/// initialization.
pub trait Component: Ident + AsComponent {
    /// Return the code for the type of the component if needed, else, return `None`.
    fn ty(&self) -> Result<proc_macro2::TokenStream, crate::Error> {
        Err(crate::Error::CodeNotProvided)
    }

    /// Return the code for the initialization expression of the component if needed, else, return `None`.
    fn init_expr(&self) -> Result<proc_macro2::TokenStream, crate::Error> {
        Err(crate::Error::CodeNotProvided)
    }

    /// Return the list of dependencies for the initialization of the component if it has any,
    /// else, return `None`.
    fn dependencies(&self) -> Option<Vec<Rc<dyn Component>>> {
        None
    }
    /// Check if the current component is a peripheral.
    ///
    /// This is needed to avoid the need for multiple structs with
    /// cyclic dependencies to the `Peripherals` "parent" component.
    fn is_peripheral(&self) -> bool {
        false
    }

    /// Return code expression that must be run after the initialization
    /// of this component if it exists.
    fn after_init(&self) -> Option<proc_macro2::TokenStream> {
        None
    }

    /// Return code expression that must be run before the initialization
    /// of this component if it exists.
    fn before_init(&self) -> Option<proc_macro2::TokenStream> {
        None
    }

    /// Return peripheral prelude code before using it inside a capsule.
    ///  TODO: This should be moved (?) in a different `Peripheral` trait,
    /// similar to `Capsule`.
    fn before_usage(&self) -> Option<proc_macro2::TokenStream> {
        None
    }
}

// Used for finding types in a list of `Component` trait objects.
impl dyn Component {
    pub fn is<T: 'static>(&self) -> bool {
        TypeId::of::<T>() == self.type_id()
    }

    pub fn downcast<T: 'static>(self: Rc<Self>) -> Result<Rc<T>, Rc<Self>> {
        if self.is::<T>() {
            unsafe { Ok(Rc::from_raw(Rc::into_raw(self) as _)) }
        } else {
            Err(self)
        }
    }
}

pub trait AsComponent {
    fn as_component(self: Rc<Self>) -> Rc<dyn Component>;
}

impl<C: Component + 'static> AsComponent for C {
    fn as_component(self: Rc<Self>) -> Rc<dyn Component> {
        self
    }
}

pub trait FormatIdent
where
    String: From<Self>,
    Self: Sized,
{
    fn format_ident(self) -> String {
        String::from(self).to_lowercase().replace('-', "_")
    }
}

impl FormatIdent for uuid::Uuid {}
