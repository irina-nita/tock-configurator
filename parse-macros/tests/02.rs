use parse_macros::component;

#[component(ident = "uart")]
pub enum Uart2 {}

fn main() {}
