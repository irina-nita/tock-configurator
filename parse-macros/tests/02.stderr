error: expected `struct`
 --> ./tests/02.rs:4:5
  |
4 | pub enum Uart2 {}
  |     ^^^^
